go/runtime/host/tdx/qemu.go

```diff
@@ -0,0 +1,361 @@
+	"github.com/oasisprotocol/oasis-core/go/runtime/host/protocol"
+	"github.com/oasisprotocol/oasis-core/go/runtime/host/sandbox"
+	"github.com/oasisprotocol/oasis-core/go/runtime/host/sandbox/process"
+	sgxCommon "github.com/oasisprotocol/oasis-core/go/runtime/host/sgx/common"
+)
+
+const (
+	// defaultQemuSystemPath is the default QEMU system binary path.
+	defaultQemuSystemPath = "/usr/bin/qemu-system-x86_64"
+	// defaultStartCid is the default start CID.
+	defaultStartCid = 0xA5150000
+	// defaultRuntimeAttestInterval is the default runtime (re-)attestation interval.
+	defaultRuntimeAttestInterval = 2 * time.Hour
+
+	// vsockPortRHP is the VSOCK port used for the Runtime-Host Protocol.
+	vsockPortRHP = 1
+
+	runtimeConnectTimeout = 30 * time.Second
+)
+
+// QemuConfig is the configuration of the QEMU-based TDX runtime provisioner.
+type QemuConfig struct {
+	// HostInfo provides information about the host environment.
+	HostInfo *protocol.HostInfo
+
+	// CommonStore is a handle to the node's common persistent store.
+	CommonStore *persistent.CommonStore
+
+	// PCS is the Intel Provisioning Certification Service client.
+	PCS pcs.Client
+	// Consensus is the consensus layer backend.
+	Consensus consensus.Backend
+	// Identity is the node identity.
+	Identity *identity.Identity
+
+	// RuntimeAttestInterval is the interval for periodic runtime re-attestation. If not specified
+	// a default will be used.
+	RuntimeAttestInterval time.Duration
+}
+
+// QemuExtraConfig is the per-runtime QEMU-specific extra configuration.
+type QemuExtraConfig struct {
+	// CID is the VSOCK cid to use for this runtime. If zero, the CID is automatically assigned.
+	CID uint32
+}
+
+type qemuProvisioner struct {
+	sync.Mutex
+
+	cfg QemuConfig
+
+	sandbox   host.Provisioner
+	pcs       pcs.Client
+	qs        pcs.QuoteService
+	consensus consensus.Backend
+	identity  *identity.Identity
+
+	nextCid uint32
+
+	logger *logging.Logger
+}
+
+// Implements host.Provisioner.
+func (q *qemuProvisioner) NewRuntime(cfg host.Config) (host.Runtime, error) {
+	q.Lock()
+	defer q.Unlock()
+
+	// Assign CID if not explicitly configured.
+	if cfg.Extra == nil {
+		cfg.Extra = &QemuExtraConfig{
+			CID: q.nextCid,
+		}
+		q.nextCid++
+	}
+	return q.sandbox.NewRuntime(cfg)
+}
+
+// Implements host.Provisioner.
+func (q *qemuProvisioner) Name() string {
+	return "tdx-qemu"
+}
+
+func (q *qemuProvisioner) getSandboxConfig(rtCfg host.Config, conn sandbox.Connector, _ string) (process.Config, error) {
+	comp, err := rtCfg.GetComponent()
+	if err != nil {
+		return process.Config{}, err
+	}
+	if comp.TEEKind() != component.TEEKindTDX {
+		return process.Config{}, fmt.Errorf("component '%s' is not a TDX component", comp.ID())
+	}
+
+	cid := rtCfg.Extra.(*QemuExtraConfig).CID
+
+	// Configure the CID on the VSOCK connector.
+	vs, ok := conn.(*vsockConnector)
+	if !ok {
+		return process.Config{}, fmt.Errorf("VSOCK connector is required")
+	}
+	vs.cid = cid
```

Comment:
I will change this so that `connector.Configure` (renamed from `ConfigureSandbox`) will also get `*host.Config` and then it can perform this in the connector which makes more sense as `Configure` must be called before `Connect`.

Reply:
Should be fixed now, please take a look.